# Lab Assignment 2 - Andrew Muncaster

**Matric No.:** 200018778 

https://github.com/ATMuncaster/PY4SA23_Assignment200018778.git

# Python Basics

Task 1

In [ ]:
# Create an If...Else statement that will test whether a number is divisible by three.

x = 9
if x % 3 == 0:
    print (str(x) + " " + "is divisable by 3")
else: print(str(x) + " " + "is not divisable by 3")

Task 2

In [ ]:
#Create an If...Else statement that will test whether a type of fruit, represented as a text string, is in a list of acceptable fruits

answer = str(input("Name of fruit:"))
if answer in ["apple", "orange", "pear", "kiwi", "strawberry"]:
    print (answer + " " + "is acceptable")
else:
    print (answer + " " + "is not acceptable")

Task 3

In [ ]:
#Create a function to calculate the distance between two coordinates using the haversine formula.
#Write the following formula where the input parameters are a pair of coordinates as two lists.
import math

def distance (lat1, lon1, lat2, lon2):
    R = 6371
    a = ((math.sin((lat1 - lat2) / 2)) ** 2) + math.cos(lat1) * math.cos(lat2) * ((math.sin((lon1 - lon2) / 2)) ** 2)
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    d = R * c
    return d

 ________________________________________________________________________________________

# Pandas and NumPy

Question 1

In [ ]:
# How many trees are of the Quercus or Acer genus? 
import pandas as pd

df = pd.read_csv('C:\\Users\\andre\\Documents\\Work\\GG3209GIS\\Assignment2\\PY4SA23_Assignment200018778\\PY4SA23_Assignment200018778\portland_park_trees.csv')
just_quercus = df[df["Genus"]=="Quercus"]
just_acer = df[df["Genus"]=="Acer"]
print(len(just_acer) + len(just_quercus))

Question 2

In [ ]:
# How many trees are of the Quercus or Acer genus and have a DBH larger than 50 inches?
import pandas as pd

dbh_quercus = df.query('Genus=="Quercus" and DBH > 50')[["Genus", "DBH"]]
print(len(dbh_quercus))

dbh_acer = df.query('Genus=="Acer" and DBH >50')[["Genus", "DBH"]]
print(len(dbh_acer))

Question 3

In [ ]:
# Which genus has the highest mean DBH of the following genera: Quercus, Acer, or Fraxinus? 
import pandas as pd

quercus_dbh = df.query('Genus=="Quercus" and DBH')
acer_dbh = df.query('Genus=="Acer" and DBH')
fraxinus_dbh = df.query('Genus=="Fraxinus" and DBH')
print(quercus_dbh["DBH"].mean())
print(acer_dbh["DBH"].mean())
print(fraxinus_dbh["DBH"].mean())

Question 4

In [ ]:
## Q4: How many different species of trees are recorded in the Acer genus?
import pandas as pd

acer_genus = df.query('Genus=="Acer"')
acer_spec = acer_genus['Genus_spec']
acer_spec.nunique()

Question 5

In [ ]:
# The world_cities_csv was imported as a pandas dataframe:
import pandas as pd
cities_df = pd.read_csv('C:\\Users\\andre\\Documents\\Work\\GG3209GIS\\Assignment2\\PY4SA23_Assignment200018778\\PY4SA23_Assignment200018778\world_cities.csv')
print(cities_df)

Calculate a new column named "pop_M" (population in millions), by transforming the
"pop" (population) column.

In [ ]:
# Calculate a new column named "pop_M" (population in millions), by transforming the "pop" (population) column.
cities_df['popM'] = cities_df['pop'] / 100000
print(cities_df)

The new popM column has been created showing the population in millions.

In [ ]:
# Remove the original "pop" column
cities_df.pop('pop')
cities_df

The original pop column has been deleted.

In [ ]:
# Choose/subset a city that starts with the same letter as your first name
import pandas as pd
import numpy as np

citiesonly_df = cities_df['city']
citiesa = np.array([c for c in citiesonly_df if c[0] == "A"])
print(citiesa[0:100])

I can now see a selection of cities to choose from. The city that I have chosen is **Aberdeen**.

In [330]:
# Subset the five biggest (i.e., largest population sizes) cities from the country where your selected city is.
cities_df.query('city=="Aberdeen"')

city country    lat  lon  capital     popM
208  Aberdeen      UK  57.15 -2.1        0  1.84031

Aberdeen belongs to the country 'UK'. The UK cities were taken from the dataframe and other columns not relating to city or population were deleted.

In [ ]:
uk_cities_df = cities_df[cities_df['country'] == "UK"]
uk_cities_df.pop('lat')
uk_cities_df.pop('lon')
uk_cities_df

The data was then sorted and the largest 5 populations subsetted.

In [411]:
uk_cities_df.sort_values(by=['popM'], ascending=False).head(5)

city country  capital      popM
21343      London      UK        1  74.89022
4502   Birmingham      UK        0   9.86969
12618     Glasgow      UK        0   6.07192
21164   Liverpool      UK        0   4.68584
20526       Leeds      UK        0   4.57875

________________________________________________________________________________________

# Python Data Visualisation

Graph 1

In [ ]:
#Graph 1: Create a scatterplot for just trees in the Ulmus genus with DBH mapped to the x-axis and tree height mapped to the y-axis
#(Hint: You will need to use the “Genus”, “DBH”, and “TreeHeight” attributes.). 
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
df = pd.read_csv('C:\\Users\\andre\\Documents\\Work\\GG3209GIS\\Assignment2\\PY4SA23_Assignment200018778\\PY4SA23_Assignment200018778\portland_park_trees.csv')
ulmus_df = df[df["Genus"]=="Ulmus"]

sns.set(style="whitegrid")
sns.set(rc={'figure.figsize':(7,7)})
sns.scatterplot(x="DBH", y="TreeHeight", data=ulmus_df)



Graph 2

In [ ]:
# Create a scatterplot for just trees in the Ulmus genus with DBH mapped to the x-axis, tree height mapped to the y-axis, and tree species mapped to hue.

sns.set(rc={'figure.figsize':(15,12)})
sns.scatterplot(x="DBH", y="TreeHeight", hue="Genus_spec", data=ulmus_df)
plt.title("Ulmus Tree Height and Diameter by Species")

Graph 3

In [ ]:
# Create a boxplot of DBH for just the Ulmus genus differentiated by species
sns.set(rc={'figure.figsize':(20,15)})
sns.boxplot(x="DBH", y="Genus_spec", hue="Genus_spec", data=ulmus_df)
plt.title("Diamater at Breast Height (DBH) by Ulmus Species")

Graph 4

In [ ]:
# Combine Graphs 1 and 3 into a single figure. Do not plot a legend for any of the graphs.

plt.rcParams['figure.figsize'] = [20, 12]
fig, axs = plt.subplots(1,2)
axs[0].scatter(x="DBH", y="TreeHeight", data=ulmus_df)
axs[1].boxplot(x="DBH", y="Genus_spec", hue="Genus_spec", data=ulmus_df)
plt.show()

____________________________________________________

# Python GeoPandas

Task 1

**Selected Dataset** - Ancient Woodland Inventory (Scotland)

In [ ]:
# Read the selected dataset as GeoPandas DataFrame
awiscot = gpd.read_file('PY4SA23_Assignment200018778/AWI_SCOTLAND_ESRI/AWI_SCOTLAND.shp')
awiscot.plot()

Task 2

In [ ]:
#Use the correct code to plot the first 5 and the last 5 sets of records in your selected dataset.
fig, ax = plt.subplots(nrows=1, ncols=2, figsize=(20, 20))

awiscot.head(5).plot(ax=ax[0])
ax[0].set_title('First 5 records')

awiscot.tail(5).plot(ax=ax[1])
ax[1].set_title('Last 5 records')

plt.show()

Task 3

In [ ]:
# Create a map where you can explore the selected dataset. Try to plot the map using some categorical attribute. Include a ToolTip.
awiscot.explore(cmap='RdYlBu')

Task 4

In [ ]:
#What is the Coordinate Reference System of the selected dataset?
awiscot.crs

The coordinate reference system of the dataset is **EPSG:27700**.

Task 5

In [ ]:
# How many features does the selected dataset contain?
awiscot

There are 17 features in this dataset.

Task 6

In [ ]:
# Define a sub-setting criterion to create a new geopandas dataframe where you filter the selected dataset based on a categorical attribute.

awiscot_lochaber = awiscot['DISTRICT'] == 'LOCHABER'
Lochaber_GDF = awiscot[awiscot_lochaber]
Lochaber_GDF

Task 7

In [ ]:
#Define a sub-setting criterion to create a new geopandas dataframe where you filter the selected dataset based on a numerical attribute.

awiscot_large = awiscot_lochaber = awiscot['HECTARE'] > 100

size_GDF = awiscot[awiscot_large]
size_GDF

Task 8

In [ ]:
#Plot the new/filtered geopandas dataframe using one of the attributes to create a choropleth map.
fig, ax = plt.subplots(1, 1)
size_GDF.plot(column='HECTARE', ax=ax, legend=True)

_______________________

# Python Rasterio

In [21]:
pwd

'C:\\Users\\andre\\Documents\\Work\\GG3209GIS\\Assignment2\\PY4SA23_Assignment200018778\\PY4SA23_Assignment200018778'

Task 1

In [22]:
# Task 1. Read the file as a rasterio dataset
import rasterio as rio
elev = rio.open('C:\\Users\\andre\\Documents\\Work\\GG3209GIS\\Assignment2\\PY4SA23_Assignment200018778\\PY4SA23_Assignment200018778\elev.tif')
elev_arr = elev.read(1)
elev_arr

array([[-32768, -32768, -32768, ..., -32768, -32768, -32768],
       [-32768, -32768, -32768, ..., -32768, -32768, -32768],
       [-32768, -32768, -32768, ..., -32768, -32768, -32768],
       ...,
       [-32768, -32768, -32768, ..., -32768, -32768, -32768],
       [-32768, -32768, -32768, ..., -32768, -32768, -32768],
       [-32768, -32768, -32768, ..., -32768, -32768, -32768]], dtype=int16)

Task 2

In [23]:
# What is the CRS of the dataset?

elev.crs

CRS.from_epsg(32617)

The coordinate reference system for elev.tif is **EPSG:32617**.

Task 3

In [26]:
# Describe the raster dataset regarding the raster extent (bounds), the reference system, and how many bands are in this dataset.

print(elev.count)
print(elev.width)
print(elev.height)

1
6303
5897


In [645]:
elev.bounds

BoundingBox(left=479753.39945587853, bottom=4170823.2037591375, right=668843.3994558785, top=4347733.203759138)

The bounds of the raster file are left - 479753.39945587853, bottom - 4170823.2037591375, right - 668843.3994558785, top - 4347733.203759138. The reference system is EPSG:32617. There is one band in this dataset.


Task 4

In [ ]:
# Create a plot/map of the raster dataset.
import os
os.environ['USE_PYGEOS'] = '0'
import geopandas
import earthpy as et
import earthpy.spatial as es
import earthpy.plot as ep
ep.plot_bands(elev_arr,
              title="Digital Elevation Model (DEM)",
              cmap="gist_earth")

plt.show()

Task 5

In [ ]:
# Create Histograms from the raster
colors_list = [
    "midnightblue",
]
titles = ["Ultra Blue"]

ep.hist(elev_arr, colors=colors_list, title=titles)
plt.show()
elev.close()